core/domain/state_domain.py

```diff
@@ -3038,11 +3053,10 @@ def to_dict(self) -> TranslatableItemDict:
 
 AcceptableConversionFnType = Union[
     Callable[[SubtitledHtml, str], SubtitledHtml],
-    Callable[
-        [Union[SubtitledHtml, SubtitledUnicode], str],
-        Union[SubtitledHtmlDict, SubtitledUnicodeDict]
-    ],
-    Callable[[Dict[str, str], str], Union[SubtitledHtml, SubtitledUnicode]],
+    Callable[[SubtitledHtml, str], SubtitledHtmlDict],
```

Comment:
I am a bit worried about bundling all these together in one type -- e.g. the first two of these seem clearly different.

What is going on here conceptually -- can you explain? There may be multiple different types of conversion functions acting on different things. What is the rationale for grouping them all together?

Reply:
If we see line 1455, then there this type is used. Basically, function `traverse_by_schema_and_convert` can accept different type of functions. So instead of Any, we decided to go with the Union of all the allowed functions that this `traverse_by_schema_and_convert` can take.

Reply:
After reading through traverse_by_schema_and_convert(), I don't think I buy that. I think it's fine to take the Union but I'm not convinced that the union you have here is correct.

From reading the function it looks to me like it deals entirely with "plain Python constructs" (dicts, strings, etc.). So why do you have domain objects like SubtitledHtml, etc. here?

Reply:
Here I think it accepts function signatures

Reply:
Here I think it accepts function signatures 
https://github.com/oppia/oppia/blob/4ba04d87b7e8aea3953287ae183c8e7d1c0031b6/core/domain/state_domain.py#L1363
and `AcceptableConversionFnType` is Union of all those function that `conversion_fn` attribute can take.
https://github.com/oppia/oppia/blob/4ba04d87b7e8aea3953287ae183c8e7d1c0031b6/core/domain/state_domain.py#L1095
https://github.com/oppia/oppia/blob/4ba04d87b7e8aea3953287ae183c8e7d1c0031b6/core/domain/state_domain.py#L1225
https://github.com/oppia/oppia/blob/4ba04d87b7e8aea3953287ae183c8e7d1c0031b6/core/domain/state_domain.py#L1273

Reply:
Aha, I know what the issue is here. I think there are functions here which convert from "customization arg that's all dict-ified" to "customization arg that uses only domain objects", and vice versa. Unfortunately they look the same outwardly because the customization args in both cases just looks like a dictionary and the keys are similar. But there are really two types of things being represented here.

I think the correct way to handle this is to define two types -- one for customization args which use domain objects for their inner values, and the other for customization args which use dicts throughout. That would bring clarity to which is which.

Reply:
Callable types are separated into two subtypes -- `DictCustomizationArgsConversionFnTypes` and `DomainObjectCustomizationArgsConversionFnTypes`.